### ISSUE TITLE
Use resolve_reference plugins in more places

### ISSUE THREAD

**Issue by adreyer** [puppetlabs/bolt#1147]
## Use Cases
* I want to use a plugin to load all inventory data
* I want to use the `pkcs7` or `prompt` plugin to load the `password` for the `vault` plugin
* I want to encrypt `vars` on a target or group.

## Out of scope
* A `resolve_reference[s]` function that calls a lookup plugin directly from a plan.

## Questions
* Do we need to limit the depth of a plugin returning a plugin?

## Prerequisites
This ticket assumes that `inventory_config` and `inventory_target` lookup hooks have been consolidated into a single `resolve_reference` hook.

## Where lookup plugins can be used
Any value in `inventory.yaml`
Items under `plugin_hooks` or `plugins` in bolt.yaml. 

## Where lookup plugins cannot be used
As a key in a yaml object anywhere in an inventoryfile/bolt.yaml
For `bolt.yaml`, only allow under an item in `plugin_hooks` and `plugins`. Disallow any other places.

## When plugins are evaluated
* When `bolt.yaml` is evaluated all references are resolved.
* If `bolt.yaml` config for one plugin references another that should happen before the first plugin is initialized. A helpful error should be raised if there is a dependency loop.
* A reference at the top level of inventory(the `all` group) is evaluated when inventory is loaded
* A reference at the top level of `groups` is evaluated when inventory is loaded
* A reference at the top level of `targets` is evaluated when inventory is loaded.
* A reference for a `name` or `uri` key should be evaluated when inventory is loaded
* A reference in `config`, `facts`, `vars` or `features` should be evaluated when it would be used by a `target`, specifically any time a `get_target(s)` call gets the target (lazily, but in a single step).
* Merging two references should merge the results of resolving the references.
* A `_plugin` object inside a second `_plugin` object should be evaluated immediately before the first `_plugin` object.
* A `_plugin` object returned from another reference should be evaluated following the above rules. ie if lookup at the top level of targets returns a `_plugin` inside a targets `config` it will be lazily loaded but if it just returns a `_plugin` object it would be immediately evaluated.

## Associated Changes
* The `targets`, `groups` and `features` array, should be flattened to generalize the existing behavior of a plugin that returns a list of targets or a single target.

**Comment by lucywyman** [comment]
Closed by #1340
